### PR TITLE
Spring I/O Spring REST Docs postblog

### PR DESCRIPTION
**Pull request by KevinDaHub**

### CODE REVIEW

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+<img class="p-image float-image" width="200" alt="Andy Wilkinson" src="/img/rest-docs/andywilkinson.jpg">
+
+### Andy Wilkinson
+
+Andy is a Spring Boot, REST docs committer and Spring IO platform lead at Pivotal. 
+You can find him on Twitter using the handle [@ankinson](https://twitter.com/ankinson).
+
+<blockquote class="clear"><p>
+Writing documentation is critical in the world of development. 
+It is used to make a accurate and straight declaration and intent of what the service has to offer.
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: A**n** accurate

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+### Andy Wilkinson
+
+Andy is a Spring Boot, REST docs committer and Spring IO platform lead at Pivotal. 
+You can find him on Twitter using the handle [@ankinson](https://twitter.com/ankinson).
+
+<blockquote class="clear"><p>
+Writing documentation is critical in the world of development. 
+It is used to make a accurate and straight declaration and intent of what the service has to offer. 
+Front-End developers will be able to know which endpoints they have to call and receive the right resource. 
+Now we all know it's tedious for developers to write documentation...
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Would add a comma:
Now**,** we all know ...

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+Price your self lucky! Spring Rest Docs will make your life easier.
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Nederengels :-)
Would replace it with something else, maybe:
"It's your lucky day!"

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+While you are writing tests, Spring will generate a fully html api guide for you and your team.
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Capitalize HTML?

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+Now we all know it's tedious for developers to write documentation...
+Price your self lucky! Spring Rest Docs will make your life easier.
+While you are writing tests, Spring will generate a fully html api guide for you and your team. 
+This blog post will take you through the best practices, how to and new features in 1.1.
+</p></blockquote>
+
+**Why Test driven approach**
+
+* it’s an accurate definition of your application (no side effects)
+* it describes the specific http request and response.
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Capitalize HTML?

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+* it’s straight forward with no repetition
+* it’s easier to write (no annotations like Swagger)
+
+### Choose your tool of use.
+
+#### Asciidoctor
+Asciidoctor is a markup language that processes plain text and produces HTML, styles and layed out to suit your needs.
+If you are interested writing in Asciidoctor be sure to checkout the [manual](http://asciidoctor.org/docs/user-manual/#introduction-to-asciidoctor).
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: If you are interested **in** writing in Asciidoctor

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+
+#### Markdown (New in 1.1)
+With the newest version of REST docs, they decided to give the developers more choice in markup languages writing the document.
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Would slightly rephrase this:
... choice in markup languages in which to write the document.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+The Markdown support is not that capable as Asciidoctor, but Markdown can work very well when combined with existing documentation toolchains such as [Slate](https://github.com/tripit/slate).
+Here is a good [sample](https://github.com/spring-projects/spring-restdocs/tree/master/samples/rest-notes-slate) working with slate.
+
+#### Andy's pick
+Asciidoctor!
+Since Asciidoctor has support for more functionality than Markdown, it gives Asciidoctor the advantage to be the preferred tool.
+
+### Test Tools
+When we want to use Spring Rest Docs, we'll have to use one of the test tools Spring has to offer. 
+Here are the different tools of choice. To use these tools we'll have to initialise which document, Mockmvc and objectWriter we'll be using.
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: You can wrap inline code snippets and variables or classes in backticks.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+tags: [Spring IO,Spring,Conference]
+category: conference
+comments: true
+---
+
+## Spring IO 2016
+The main focus this year was definitely about Cloud, Reactive and MicroServices.
+But to complete this story we will need to document the REST API. 
+Keep calm, you don't have to do it manually! 
+Spring made it easy for us with Spring Rest Docs!
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Would replace all "Spring Rest Docs" with "Spring **REST** Docs".

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+                                fieldWithPath("_links").description("links to other resources")
+                        ));
+ mockMvc.perform(
+                get("/employees/1").accept(MediaType.APPLICATION_JSON));
+
+{% endhighlight %}
+
+
+#### Relaxed Snippet (New in 1.1)
+In contrast of default snippets, relaxed snippets are used to have no complaining when neglected something in the document.
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: In contrast **to**

Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: ... when **something was neglected** in ...

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+</p></blockquote>
+
+**Why Test driven approach**
+
+* it’s an accurate definition of your application (no side effects)
+* it describes the specific http request and response.
+* it’s straight forward with no repetition
+* it’s easier to write (no annotations like Swagger)
+
+### Choose your tool of use.
```
Comment (tomverelst) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: tool of choice?

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+                        fieldWithPath("username").description("The employee unique database identifier").type(String.class),
+                        fieldWithPath("_links").description("links to other resources")
+                ));
+{% endhighlight %}
+
+
+### Request fields
+This defines your fields you request from the client.
+Normally it will complain if you neglected a field but with v1.1 Spring REST Docs has support for Relaxed Snippets.
+Since I use constraints I made my own method withPath, this will add an extra column constraint to the documentation.
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Would wrap withPath in backticks:
`withPath`

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+title: 'Spring I/O 16: Test-driven documentation with Spring REST Docs'
+image: /img/springio.jpg
+tags: [Spring IO,Spring,Conference]
+category: conference
+comments: true
+---
+
+## Spring IO 2016
+The main focus this year was definitely about Cloud, Reactive and MicroServices.
+But to complete this story we will need to document the REST API.
```
Comment (tomverelst) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: We need to document **the** REST API? We were talking about this year's focus and now I have to document **the** REST API :)?

I'd rephrase this to, "but it is important not to forget other topics, like documentation, ..." or something similar.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+                       fieldWithPath("username").description("The employee unique database identifier"),
+                       fieldWithPath("firstName").description("The employee's first name"),
+                       fieldWithPath("lastName").description("The employee's last name"),
+                           ));
+{% endhighlight %}
+
+
+### Request/response headers
+Defines your request/response headers in your api. 
+This is useful when there are extra headers to use.
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: "to use" -> "to **set**" or "to **specify**"?

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+
+
+
+
+### 3 main questions if you are working with resources
+   * What do they represent?
+   * What input do they accept?
+   * What output do they produce?
+
+Last but not least: dont document uri’s!
```
Comment (tomverelst) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: do not

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+
+## Questions
+
+### Will constraints be officially added in future releases?
+The constraints snippets won't be added in the future.
```
Comment (tomverelst) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: constraint's

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+			parameterWithName("page").description("The page to retrieve"), 
+			parameterWithName("per_page").description("Entries per page") 
+	)));
+	
+	
+{% endhighlight %}
+
+
+### Request parts (new in 1.1)
+The parts of a multipart request can be documenting using requestParts
```
Comment (yannickdeturck) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Would wrap requestParts in backticks.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+authors: [kevin_van_houtte]
+title: 'Spring I/O 16: Test-driven documentation with Spring REST Docs'
+image: /img/springio.jpg
+tags: [Spring IO,Spring,Conference]
+category: conference
+comments: true
+---
+
+## Spring IO 2016
+The main focus this year was definitely about Cloud, Reactive and MicroServices.
```
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: microservices instead of MicroServices

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+But it is important not to forget other topics, like documentation! 
+Keep calm, you don't have to do it manually! 
+Spring made it easy for us with Spring REST Docs! 
+This year in Spring IO, Andy Wilkinson himself talked about why, how and when Spring Rest Docs are being used.
```
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: This year AT Spring IO

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md (hunk 5 above)
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: No dot at the end

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+</p></blockquote>
+
+**Why Test driven approach**
+
+* it’s an accurate definition of your application (no side effects)
+* it describes the specific http request and response.
+* it’s straight forward with no repetition
+* it’s easier to write (no annotations like Swagger)
+
+### Choose your tool of choice.
```
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: No dot at the end

Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Choose ... choice -> duplication
Better: Pick your tool of choice

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+
+#### Asciidoctor
+Asciidoctor is a markup language that processes plain text and produces HTML, styles and layed out to suit your needs.
```
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: and produces HTML, completely styled to suit your needs.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 22]
```diff
@@ -0,0 +1,339 @@
...
+If you are interested in writing in Asciidoctor be sure to checkout the [manual](http://asciidoctor.org/docs/user-manual/#introduction-to-asciidoctor).
```
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: be sure to check out the

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 23]
```diff
@@ -0,0 +1,339 @@
...
+
+#### Markdown (New in 1.1)
+With the newest version of REST docs, they decided to give the developers more choice in markup languages to write the document.
```
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: REST Docs, the developer now has more options in terms of markup languages.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 24]
```diff
@@ -0,0 +1,339 @@
...
+The Markdown support is not that capable as Asciidoctor, but Markdown can work very well when combined with existing documentation toolchains such as [Slate](https://github.com/tripit/slate).
```
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: is not as feature-rich as Asciidoctor,

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 25]
```diff
@@ -0,0 +1,339 @@
...
+Here is a good [sample](https://github.com/spring-projects/spring-restdocs/tree/master/samples/rest-notes-slate) working with slate.
+
+#### Andy's pick
+Asciidoctor!
+Since Asciidoctor has support for more functionality than Markdown, it gives Asciidoctor the advantage to be the preferred tool.
```
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Since Asciidoctor boasts more features than Markdown, it gives Asciidoctor the edge.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 26]
```diff
@@ -0,0 +1,339 @@
...
+
+### Test Tools
+When we want to use Spring Rest Docs, we'll have to use one of the test tools Spring has to offer.
```
Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Spring REST Docs

Comment (andreasevers) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: We don't have to use a Spring tool, do we?
RestAssured isn't from Spring afaik.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 27]
```diff
@@ -0,0 +1,339 @@
...
+
+### Andy Wilkinson
+
+Andy is a Spring Boot, REST docs committer and Spring IO platform lead at Pivotal. 
+You can find him on Twitter using the handle [@ankinson](https://twitter.com/ankinson).
+
+<blockquote class="clear"><p>
+Writing documentation is critical in the world of development. 
+It is used to make an accurate and straight declaration and intent of what the service has to offer. 
+Front-End developers will be able to know which endpoints they have to call and receive the right resource.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: replace 'resource' with 'data' (or 'information').

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md (hunk 5 above)
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: capital letter at beginning of each line

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 28]
```diff
@@ -0,0 +1,339 @@
...
+It's your lucky day! Spring Rest Docs will make your life easier.
+While you are writing tests, Spring will generate a fully HTML api guide for you and your team. 
+This blog post will take you through the best practices, how to and new features in 1.1.
+</p></blockquote>
+
+**Why Test driven approach**
+
+* it’s an accurate definition of your application (no side effects)
+* it describes the specific http request and response.
+* it’s straight forward with no repetition
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: with  no --> without repetition

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 29]
```diff
@@ -0,0 +1,339 @@
...
+When we want to use Spring Rest Docs, we'll have to use one of the test tools Spring has to offer. 
+Here are the different tools of choice. To use these tools we'll have to initialise which document, `Mockmvc` and `ObjectWriter` we'll be using. 
+For security there is support for adding filters.
+
+#### MockMvc
+A lightweight server-less documentation generation by the Spring Framework that has been the default use in Spring REST Docs.
+
+
+{% highlight java %}
+private MockMvc mockMvc;
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: should there be "MockMvc" twice?

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 30]
```diff
@@ -0,0 +1,339 @@
...
+                                fieldWithPath("_links").description("links to other resources")
+                        ));
+ mockMvc.perform(
+                get("/employees/1").accept(MediaType.APPLICATION_JSON));
+
+{% endhighlight %}
+
+
+#### Relaxed Snippet (New in 1.1)
+In contrast to default snippets, relaxed snippets are used to have no complaining when something was neglected in the document.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: ... relaxed snippets don't complain when something was ...

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 31]
```diff
@@ -0,0 +1,339 @@
...
+This has the advantage if you would like to focus on a certain scenario or specific part of the response/request.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: This is an advantage when you only need to focus on a certain scenario or specific part ...

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 32]
```diff
@@ -0,0 +1,339 @@
...
+
+#### Reusable Snippet (New in 1.1)
+With the newly introduced reusable snippet, you can define a snippet in the beginning of your test class and reuse it every time you need it.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: at the beginning of your testclass

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 33]
```diff
@@ -0,0 +1,339 @@
...
+            linkWithRel("alpha").description("Link to the alpha resource"),
+            linkWithRel("bravo").description("Link to the bravo resource"))));
+{% endhighlight %}
+
+
+## Type of Snippets:
+A snippet can be one of the following:
+
+### Hypermedia links
+When documenting your hypermedia application, you'll have to define your links and what they link to.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: your links and where they go.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 34]
```diff
@@ -0,0 +1,339 @@
...
+                        linkWithRel("employee").optional().description("The employee's projection")),
+                responseFields(
+                        fieldWithPath("username").description("The employee unique database identifier").type(String.class),
+                        fieldWithPath("_links").description("links to other resources")
+                ));
+{% endhighlight %}
+
+
+### Request fields
+This defines your fields you request from the client.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: this defines the fields you ....

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 35]
```diff
@@ -0,0 +1,339 @@
...
+Normally it will complain if you neglected a field but with v1.1 Spring REST Docs has support for Relaxed Snippets.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: .... complain when you neglect a field but since v1.1 Spring REST Docs supports Relaxed Snippets

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 36]
```diff
@@ -0,0 +1,339 @@
...
+Since I use constraints I made my own method `withPath, this will add an extra column constraint to the documentation.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: Because I use constraints I've created my own method "withPath". This will ....

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 37]
```diff
@@ -0,0 +1,339 @@
...
+                       fieldWithPath("firstName").description("The employee's first name"),
+                       fieldWithPath("lastName").description("The employee's last name"),
+                           ));
+{% endhighlight %}
+
+
+### Request/response headers
+Defines your request/response headers in your api. 
+This is useful when there are extra headers to set. 
+If the request has to involve an authorization header for security you can add this header to your document.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: When the request has to involve an authorization header for security reasons, you can add this header to your document.

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 38]
```diff
@@ -0,0 +1,339 @@
...
+              responseFields(
+                       fieldWithPath("username").description("The employee unique database identifier"),
+                       fieldWithPath("firstName").description("The employee's first name"),
+                       fieldWithPath("lastName").description("The employee's last name"),
+                           ));
+{% endhighlight %}
+
+
+### Request/response headers
+Defines your request/response headers in your api.
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: API

File: _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md [hunk 39]
```diff
@@ -0,0 +1,339 @@
...
+#### And HTTP verbs/codes (PATCH VS PUT)
+To be RESTfull, you'll have to follow the guidelines in having a correct API design. 
+This [picture](https://upload.wikimedia.org/wikipedia/commons/8/88/Http-headers-status.png) shows you how and when to use the correct verbs and HTTP codes
+
+
+
+
+### 3 main questions if you are working with resources
+   * What do they represent?
+   * What input do they accept?
```
Comment (kabr) [inline _posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md]: What kind of input do they accept?
